### PR TITLE
Add new datastore backend

### PR DESCRIPTION
**Pull request by mrkwse**
[See Confluence for further context](https://confluence.dev.bbc.co.uk/display/rdprodenv/NMOS+Registry+Switchover)

Add parallel implementation of Couchbase document store backend alongside etcd.

Still WIP, however _should_ be functionally equivalent to `etcd` backend at this point. Still a bit of refactoring/tidying, and a couple of places where behaviour still a little bit legacy.

### CODE REVIEW

File: nmosregistration/aggregation.py [hunk 1]
```diff
@@ -38,23 +39,33 @@
...
         self._config = config
 
-        garbage_collect_interval = int(self._config.get("garbage_collect_interval", 10))
-        self._garbage_collector = GarbageCollect(identifier=HOST, registry=registry, interval=garbage_collect_interval)
+        if config['registry']['type'] == 'etcd':
+            garbage_collect_interval = int(self._config.get("garbage_collect_interval", 10))
+            self._garbage_collector = GarbageCollect(identifier=HOST, registry=registry, interval=garbage_collect_interval)
 
-        self._v1_0_api = v1_0.Routes(logger=logger, registry=registry)
+        self._v1_0_api = v1_0.Routes(logger=logger, registry=registry, registry_type=config['registry']['type'])
```
Comment (andrewbonney) [inline nmosregistration/aggregation.py]: Could the type of registry be made a property of the registry object so we don't need to pass anything extra in here?

File: nmosregistration/common/routes.py [hunk 2]
```diff
@@ -24,16 +24,19 @@
 from nmosregistration.modifier import RegModifier
 from nmosregistration.common import schema
 
+from requests.models import Response
+
 VALID_TYPES = ['node', 'source', 'flow', 'device', "receiver", "sender"]
 REGISTRY_PORT = 4001
 NODE_SEEN_TTL = 12  # seconds until a node considered "dead".
```
Comment (andrewbonney) [inline nmosregistration/common/routes.py]: Obviously fine to use these as they're already here, but 'REGISTRY_PORT' and 'NODE_SEEN_TTL' (which should really be 'heartbeat_timeout' or similar) would be better derived from the config.

File: nmosregistration/couchbase_backend.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+    def upsert(self, rtype, rkey, value, xattrs, ttl=12):
+        try:
+            upsert_result = self.registry.upsert(rkey, value, ttl=ttl)
+            r = make_response(json.dumps(value), 200)
+        except couchbase.exceptions.KeyExistsError:
+            return make_response(409)
+        if upsert_result.success:
+            write_time = Timestamp.get_time().to_nanosec()
+            time.sleep(1)
```
Comment (andrewbonney) [inline nmosregistration/couchbase_backend.py]: Any reason for this sleep?

File: nmosregistration/couchbase_backend.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        try:
+            touch_result = self.registry.touch(rkey, ttl=ttl)
+        except Exception:
+            self.remove(rkey)
+            r = make_response(500)
+
+        return r
+
+    # Legacy put command warps around upsert. Sanitises inputs, removing etcd specific decoration.
+    def put(self, rtype, rkey, value, ttl=12, port=None):
```
Comment (andrewbonney) [inline nmosregistration/couchbase_backend.py]: The TTL default probably ought to come from config, and match the 'NODE_SEEN_TTL' from elsewhere.

File: nmosregistration/aggregation.py [hunk 5]
```diff
@@ -38,23 +39,33 @@
...
+            )
+        else:
+            registry = EtcdInterface()
         super(AggregatorAPI, self).__init__()
         self._config = config
 
-        garbage_collect_interval = int(self._config.get("garbage_collect_interval", 10))
-        self._garbage_collector = GarbageCollect(identifier=HOST, registry=registry, interval=garbage_collect_interval)
+        if config['registry']['type'] == 'etcd':
+            garbage_collect_interval = int(self._config.get("garbage_collect_interval", 10))
```
Comment (andrewbonney) [inline nmosregistration/aggregation.py:50]: Could this be definition be moved into the config file and the start of the service file?

File: nmosregistration/registryaggregatorservice.py [hunk 6]
```diff
@@ -46,9 +46,23 @@
...
+                "type": "etcd",
+                "hosts": ["localhost"],
+                "port": 4001,
+                "username": "",
+                "password": "",
+                "bucket": "",
+            }
+        }
         self._load_config()
+        if registry:
```
Comment (andrewbonney) [inline nmosregistration/registryaggregatorservice.py]: Is this just to aid testing? If so would patching achieve the same thing?

File: nmosregistration/couchbase_backend.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        self.bucket = bucket
+
+    class RegistryUnavailable(Exception):
+        pass
+
+    def _get_associated_node_id(self, rtype, value):
+        if rtype == 'device':
+            return value['node_id']
+        elif rtype in ['receiver', 'sender', 'source', 'flow']:
+            return self.registry.get(value['device_id']).value['node_id']
```
Comment (andrewbonney) [inline nmosregistration/couchbase_backend.py]: Watch out for v1.0 Flows which don't have a device_id.

**Review by andrewbonney — COMMENTED** [review]
Looks good so far. Just added a few comments which I think are all minor things.

File: nmosregistration/couchbase_backend.py (hunk 3 above)
Comment (mrkwse) [reply nmosregistration/couchbase_backend.py]: I think that is some debugging detritus, gone. :+1:

File: tox.ini [hunk 8]
```diff
@@ -9,24 +9,29 @@
...
+; deps =
+;     coverage
+;     mock
+;     testcontainers
 
 [testenv:py3]
 commands =
-    coverage run --source=./nmosregistration -m unittest discover -s tests
+    ; coverage run --source=./nmosregistration -m unittest discover -s tests
+    coverage run --source=./nmosregistration -m pytest
```
Comment (mrkwse) [inline tox.ini:26]: Commit should have referenced move to pytest for more reliable (blocking/non-concurrent) integration testing.

**Comment by andrewbonney** [comment]
Looks like the dependency addition in setup.py may need replicating into stdeb.cfg in a Debian/Ubuntu compatible way:

```
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]: Traceback (most recent call last):
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:   File "/usr/bin/nmosregistration", line 7, in <module>
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:     from nmosregistration.registryaggregatorservice import RegistryAggregatorService
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:   File "/usr/lib/python2.7/dist-packages/gevent/builtins.py", line 93, in __import__
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:     result = _import(*args, **kwargs)
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:   File "/usr/lib/python2.7/dist-packages/nmosregistration/registryaggregatorservice.py", line 21, in <module>
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:     from nmosregistration.aggregation import AggregatorAPI, AGGREGATOR_APIVERSIONS # noqa E402
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:   File "/usr/lib/python2.7/dist-packages/gevent/builtins.py", line 93, in __import__
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:     result = _import(*args, **kwargs)
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:   File "/usr/lib/python2.7/dist-packages/nmosregistration/aggregation.py", line 20, in <module>
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:     from .couchbase_backend import CouchbaseInterface
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:   File "/usr/lib/python2.7/dist-packages/gevent/builtins.py", line 93, in __import__
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:     result = _import(*args, **kwargs)
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:   File "/usr/lib/python2.7/dist-packages/nmosregistration/couchbase_backend.py", line 18, in <module>
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:     import couchbase.exceptions
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:   File "/usr/lib/python2.7/dist-packages/gevent/builtins.py", line 93, in __import__
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]:     result = _import(*args, **kwargs)
Aug 12 14:58:14 ap-z220-0 nmosregistration[12841]: ImportError: No module named couchbase.exceptions
Aug 12 14:58:14 ap-z220-0 systemd[1]: python-registryaggregator.service: Main process exited, code=exited, status=1/FAILURE
Aug 12 14:58:14 ap-z220-0 systemd[1]: python-registryaggregator.service: Unit entered failed state.
Aug 12 14:58:14 ap-z220-0 systemd[1]: python-registryaggregator.service: Failed with result 'exit-code'.
```

If the package is also becoming Python 3 only then the builds for Py2 packages will need to stop. I can't remember if that was a requirement or not.

**Comment by andrewbonney** [comment]
If Debianising turns out to be a pain, we may use this as a first step into using pure Ansible/PyPI for deployment.

**Pull request closed without merging**